pkg/internalregistry/controllers/service_account_controller.go

```diff
@@ -134,16 +139,56 @@ func (c *serviceAccountController) sync(ctx context.Context, key string) error {
 	// name of managed secret
 	secretName, err := c.managedImagePullSecretName(ctx, serviceAccount)
 
-	// ensure secret ref annotation is set
-	if serviceAccount.Annotations[InternalRegistryImagePullSecretRefKey] != secretName {
-		patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
-		if err != nil {
-			return err
+	// extract the apply configuration patch from the cache
+	patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
+	if err != nil {
+		return err
+	}
+
+	// we always want the annotation set
+	patch.WithAnnotations(map[string]string{InternalRegistryImagePullSecretRefKey: secretName})
+
+	// TODO stop adding to .Secrets part of API-1798
+	if !slices.ContainsFunc(patch.Secrets, func(ref applycorev1.ObjectReferenceApplyConfiguration) bool { return *ref.Name == secretName }) {
+		patch.WithSecrets(applycorev1.ObjectReference().WithName(secretName))
+	}
+
+	// We have tp be careful with the atomic ServiceAccount.ImagePullSecrets field so
+	// that we do not race with another controller or user. Add ImagePullSecrets to
+	// the apply patch explicitly only if the value we want to set it to was not the
+	// value extracted by the ExtractServiceAccount() func, and if the managed image
+	// pull secrets is not already referenced in the service account ImagePullSecrets
+	// field.
+	//  | Contains managed image pull secret ref:
+	//  |               \ |                   patch                             |
+	//  | serviceaccount \|  true               | false                         |
+	//  | --------------- | ------------------- | ----------------------------- |
+	//  | true            |  Already in patch   | Ignore (other owner exists)   |
+	//  | false           |  Should not happen  | Add to patch (take ownership) |
+	//
+	serviceAccountContainsImagePullSecretRef := slices.ContainsFunc(serviceAccount.ImagePullSecrets, func(ref corev1.LocalObjectReference) bool { return ref.Name == secretName })
+	patchContainsImagePullSecretRef := slices.ContainsFunc(patch.ImagePullSecrets, func(ref applycorev1.LocalObjectReferenceApplyConfiguration) bool { return *ref.Name == secretName })
+	if !serviceAccountContainsImagePullSecretRef && !patchContainsImagePullSecretRef {
+		patch.ImagePullSecrets = nil
+		for _, ref := range serviceAccount.ImagePullSecrets {
+			patch.WithImagePullSecrets(applycorev1.LocalObjectReference().WithName(ref.Name))
 		}
-		patch.WithAnnotations(map[string]string{InternalRegistryImagePullSecretRefKey: secretName})
+		patch.WithImagePullSecrets(applycorev1.LocalObjectReference().WithName(secretName))
+	}
+
+	// apply patch only if necessary
+	extracted, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
+	if err != nil {
+		return err
+	}
+	if !cmp.Equal(patch, extracted) {
```

Comment:
After our earlier discussion, and because the all of the generated ApplyConfiguration fields have omitempty in their JSON tag, I think we want to use `apiequality.Semantic` for this comparison.

Reply:
Done.